h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java

```diff
@@ -0,0 +1,153 @@
+    }
+
+    @Override
+    public Val apply(Env env, Env.StackHelp stk, AstRoot[] asts) {
+        Frame inputFrame = stk.track(asts[1].exec(env).getFrame());
+        boolean preprocess = asts[2].exec(env).getBool();
+
+        // Pre-processing
+        Frame wordFrame;
+        long documentsCnt;
+        if (preprocess) {
+            byte[] outputTypes = new byte[]{ Vec.T_NUM, Vec.T_STR };
+
+            TfIdfPreprocessor preprocessor = new TfIdfPreprocessor();
+            wordFrame = preprocessor.doAll(outputTypes, inputFrame).outputFrame(PREPROCESSED_FRAME_COL_NAMES, null);
+
+            documentsCnt = inputFrame.numRows();
+        } else {
+            wordFrame = inputFrame;
+            // TODO: Avoid converting to categorical Vec
+            documentsCnt = inputFrame.vec(0).toCategoricalVec().cardinality();
+        }
+
+        // DF
+        DocumentFrequency dfTaks = new DocumentFrequency();
+        Frame dfOutFrame = dfTaks.compute(wordFrame);
+
+        // IDF
+        InverseDocumentFrequency idf = new InverseDocumentFrequency(documentsCnt);
+        Vec idfValues = idf.doAll(new byte[]{ Vec.T_NUM }, dfOutFrame.lastVec()).outputFrame().anyVec();
+        // Replace DF column with IDF column
+        dfOutFrame.remove(dfOutFrame.numCols() - 1);
+        dfOutFrame.add(IDF_COL_NAME, idfValues);
+
+        // TF
+        TermFrequency tfTask = new TermFrequency();
+        Frame tfOutFrame = tfTask.compute(wordFrame);
+        
+        // Intermediate frame containing both TF and IDF values
+        // TODO: Avoid converting a string column to categorical
+        tfOutFrame.replace(1, tfOutFrame.vecs()[1].toCategoricalVec());
+        dfOutFrame.replace(0, dfOutFrame.vecs()[0].toCategoricalVec());
+        int[][] levelMaps = {
+                CategoricalWrappedVec.computeMap(tfOutFrame.vec(1).domain(), dfOutFrame.vec(0).domain())
+        };
+        Frame tfIdfIntermediate = Merge.merge(tfOutFrame, dfOutFrame, new int[]{ 1 }, new int[]{ 0 }, false, levelMaps);
+        // TODO: Convert columnt back to string
+        tfIdfIntermediate.replace(1, tfIdfIntermediate.vecs()[1].toStringVec());
+
+        // TF-IDF
+        int tfOutFrameColCnt = tfIdfIntermediate.numCols();
+        TfIdf tfIdfTask = new TfIdf(tfOutFrameColCnt - 2, tfOutFrameColCnt - 1);
+        Vec tfIdfValues = tfIdfTask.doAll(new byte[]{ Vec.T_NUM }, tfIdfIntermediate).outputFrame().anyVec();
+
+        // Construct final frame containing TF, IDF and TF-IDF values
+        tfIdfIntermediate.add(TF_IDF_COL_NAME, tfIdfValues);
+        tfIdfIntermediate._key = Key.make();
+
+        System.out.println(tfIdfIntermediate.toTwoDimTable().toString());
+        
+        return new ValFrame(tfIdfIntermediate);
+    }
+
+    @Override
+    public String str() {
+        return "tf-idf";
+    }
+
+    /**
+     * Final TF-IDF Map-Reduce task used to combine TF and IDF values together.
+     */
+    private static class TfIdf extends MRTask<TfIdf> {
```

Comment:
I would recommend to name it as:
```suggestion
    private static class TfIdfTask extends MRTask<TfIdf> {
```

The same for your other MRTasks.

Reply:
Sounds good. I renamed all of the `MRTask` in such way. Also, I renamed "tasks" which are not actual `MRTaks` - `DocumentFrequency`, `TermFrequency`... to be consistent.